### PR TITLE
Feature/air 8530

### PR DESCRIPTION
**Pull request by graemedewe**

### CODE REVIEW

File: src/ShoppingCart/Factory/ShoppingCartServiceFactory.php [hunk 1]
```diff
@@ -2,36 +2,57 @@
...
+        if (!isset($config['shopping_cart'])) {
             throw new \Exception('Configuration ShoppingCart not set.');
         }
-        
+
+        /** @var OneTimeAccess $oneTimeAccess */
+        $oneTimeAccess = $serviceLocator->get(OneTimeAccess::class);
+
+        /** @var MVoucherCoupon  $mVoucherCoupon */
+        $mVoucherCoupon = $serviceLocator->get(MVoucherCoupon::class);
```
Comment (TotalWipeOut) [inline src/ShoppingCart/Factory/ShoppingCartServiceFactory.php]: This shouldn't be a service, it has no dependencies. It should just be instantiated with `new` when its needed when the coupon details are available...

File: src/ShoppingCart/Service/Coupon/MVoucherCoupon.php [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+     */
+    private $offerDiscountType;
+
+    /**
+     * @param $code
+     * @param $details
+     *
+     * @return $this
+     */
+    public function setUp($code, $details)
```
Comment (TotalWipeOut) [inline src/ShoppingCart/Service/Coupon/MVoucherCoupon.php]: this really should be the constructor -  the object is useless without this information.

File: src/ShoppingCart/Factory/OneTimeAccessFactory.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     * @return mixed|\ShoppingCart\Service\OneTimeAccess
+     * @throws \Zend\Session\Exception\InvalidArgumentException
+     * @throws \Zend\ServiceManager\Exception\ServiceNotFoundException
+     */
+    public function createService(ServiceLocatorInterface $serviceLocator)
+    {
+        $container = new Container('ota');
+
+        /** @var \Doctrine\Common\Persistence\ObjectManager $objectManager */
+        $objectManager = $serviceLocator->get('doctrine.entitymanager.orm_default');
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Factory/OneTimeAccessFactory.php]: done

File: src/ShoppingCart/Factory/ShoppingCartPluginFactory.php [hunk 4]
```diff
@@ -2,28 +2,38 @@
...
+     */
+    public function createService(ServiceLocatorInterface $serviceLocator)
     {
-        $shoppingCartService = $servicelocator->getServiceLocator()->get('ShoppingCartService');
-        
+        $shoppingCartService = $serviceLocator->getServiceLocator()->get('ShoppingCartService');
+
+        $oneTimeAccess = $serviceLocator->getServiceLocator()->get(OneTimeAccess::class);
+
+        /** @var \Monolog\Logger $logger */
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Factory/ShoppingCartPluginFactory.php]: fixed

File: src/ShoppingCart/Factory/ShoppingCartPluginFactory.php [hunk 5]
```diff
@@ -2,28 +2,38 @@
...
-use ShoppingCart\Entity\ShoppingCartEntity;
 
 class ShoppingCartPluginFactory implements FactoryInterface
 {
-
-    public function createService(ServiceLocatorInterface $servicelocator)
+    /**
+     * @param \Zend\ServiceManager\ServiceLocatorInterface $serviceLocator
+     *
+     * @return mixed|\ShoppingCart\Controller\Plugin\ShoppingCartPlugin
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Factory/ShoppingCartPluginFactory.php]: fixed

File: src/ShoppingCart/Factory/ShoppingCartServiceFactory.php [hunk 6]
```diff
@@ -2,36 +2,57 @@
...
-use ShoppingCart\Hydrator\ShoppingCartHydrator;
-use ShoppingCart\Entity\ShoppingCartEntity;
 
 class ShoppingCartServiceFactory implements FactoryInterface
 {
-
+    /**
+     * @param \Zend\ServiceManager\ServiceLocatorInterface $serviceLocator
+     *
+     * @return mixed|\ShoppingCart\Service\ShoppingCartService
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Factory/ShoppingCartServiceFactory.php]: fixed

File: src/ShoppingCart/Service/ShoppingCartService.php [hunk 7]
```diff
@@ -1,9 +1,10 @@
 <?php
 namespace ShoppingCart\Service;
 
+use Monolog\Logger;
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Service/ShoppingCartService.php]: fixed

File: config/service.config.php [hunk 8]
```diff
@@ -1,6 +1,15 @@
...
-);
\ No newline at end of file
+
+return [
+    'factories'  => [
+        'ShoppingCartService'                     => 'ShoppingCart\Factory\ShoppingCartServiceFactory',
+        ShoppingCart\Service\OneTimeAccess::class => ShoppingCart\Factory\OneTimeAccessFactory::class,
+    ],
+    'invokables' => [
+        ShoppingCart\Service\Coupon\MVoucherCoupon::class => ShoppingCart\Service\Coupon\MVoucherCoupon::class,
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply config/service.config.php]: changed back

File: src/ShoppingCart/Service/Coupon/MVoucherCoupon.php [hunk 9]
```diff
@@ -0,0 +1,187 @@
+<?php
+namespace ShoppingCart\Service\Coupon;
+
+/**
+ * Service (was a Model) for a mVoucher coupon. The constructor was removed and now SetUp() just expects the token and
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Service/Coupon/MVoucherCoupon.php]: changed to a service so could be extended if it required dependencies in the future

File: src/ShoppingCart/Service/ShoppingCartService.php [hunk 10]
```diff
@@ -61,6 +77,99 @@ public function insert(array $items)
...
+    /**
+     * @param $clientRef
+     * @param $partnerIdentifier
+     * @param $ipAddress
+     *
+     * @return string
+     * @throws \Doctrine\ORM\OptimisticLockException
+     * @throws \ShoppingCart\Exception\OTARuntimeException
+     */
+    public function generateOta($clientRef, $partnerIdentifier, $ipAddress)
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Service/ShoppingCartService.php]: done , missed that before

File: src/ShoppingCart/Service/ShoppingCartService.php [hunk 11]
```diff
@@ -61,6 +77,99 @@ public function insert(array $items)
...
+        }
+
+        return null;
+    }
+
+    /**
+     * load from session
+     * @throws \ShoppingCart\Exception\OTARuntimeException
+     */
+    public function loadFromSession()
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Service/ShoppingCartService.php]: this is ?? refactored slightly

File: src/ShoppingCart/Service/ShoppingCartService.php [hunk 12]
```diff
@@ -61,6 +77,99 @@ public function insert(array $items)
...
 
+    /**
+     * @param $tp
+     * @param $hash
+     *
+     * @return array
+     * @throws \Doctrine\ORM\OptimisticLockException
+     * @throws \ShoppingCart\Exception\OTARuntimeException
+     */
+    public function otaDiscount($tp, $hash)
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Service/ShoppingCartService.php]: done

File: src/ShoppingCart/Controller/Plugin/ShoppingCartPlugin.php [hunk 13]
```diff
@@ -81,6 +102,53 @@ public function total_sum ($round = 2, $with_vat = false) {
...
+     * @param $partnerIdentifier
+     * @param $ipAddress
+     *
+     * @return string
+     * @throws \Doctrine\ORM\OptimisticLockException
+     * @throws \ShoppingCart\Exception\OTARuntimeException
+     */
+    public function generateOta($clientRef, $partnerIdentifier, $ipAddress)
+    {
+        return $this->getShoppingCartService()->generateOta($clientRef, $partnerIdentifier, $ipAddress);
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Controller/Plugin/ShoppingCartPlugin.php]: done

File: src/ShoppingCart/Controller/Plugin/ShoppingCartPlugin.php [hunk 14]
```diff
@@ -81,6 +102,53 @@ public function total_sum ($round = 2, $with_vat = false) {
...
+     * @param $tp
+     * @param $hash
+     *
+     * @return array
+     * @throws \Doctrine\ORM\OptimisticLockException
+     * @throws \ShoppingCart\Exception\OTARuntimeException
+     */
+    public function otaDiscount($tp, $hash)
+    {
+        return $this->getShoppingCartService()->otaDiscount($tp,$hash);
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Controller/Plugin/ShoppingCartPlugin.php]: done

File: src/ShoppingCart/Controller/Plugin/ShoppingCartPlugin.php [hunk 15]
```diff
@@ -81,6 +102,53 @@ public function total_sum ($round = 2, $with_vat = false) {
...
+    }
+
+    /**
+     * load from session
+     *
+     * @throws \ShoppingCart\Exception\OTARuntimeException
+     */
+    public function loadFromSession()
+    {
+        $this->getShoppingCartService()->loadFromSession();
```
[the remark this answers is not in the document]
Comment (graemedewe) [reply src/ShoppingCart/Controller/Plugin/ShoppingCartPlugin.php]: done

File: src/ShoppingCart/Factory/ShoppingCartServiceFactory.php (hunk 1 above)
Comment (graemedewe) [reply src/ShoppingCart/Factory/ShoppingCartServiceFactory.php]: done

File: src/ShoppingCart/Service/Coupon/MVoucherCoupon.php (hunk 2 above)
Comment (graemedewe) [reply src/ShoppingCart/Service/Coupon/MVoucherCoupon.php]: changed back

**Pull request merged**